kong/dns/client.lua

```diff
@@ -0,0 +1,761 @@
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = errcode == NAME_ERROR_CODE and self.empty_ttl or self.error_ttl
+    return answers
+  end
+
+  local processed_answers = {}
+  local cname_answer
+
+  local ttl = math_min(self.valid_ttl or 0xffffffff, 0xffffffff)  -- 0xffffffff for maximum TTL value
+
+  for _, answer in ipairs(answers) do
+    answer.name = string_lower(answer.name)
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+
+    if answer.type == TYPE_CNAME then
+      cname_answer = answer   -- use the last one as the real cname
+
+    elseif answer.type == qtype then
+      -- compatible with balancer, see https://github.com/Kong/kong/pull/3088
+      if answer.type == TYPE_AAAA then
+        answer.address = ipv6_bracket(answer.address)
+
+      elseif answer.type == TYPE_SRV then
+        answer.target = ipv6_bracket(answer.target)
+      end
+
+      -- skip the SRV record pointing to itself,
+      -- see https://github.com/Kong/lua-resty-dns-client/pull/3
+      if not (answer.type == TYPE_SRV and answer.target == qname) then
+        table_insert(processed_answers, answer)
+      end
+    end
+  end
+
+  if table_isempty(processed_answers) then
+    if not cname_answer then
+      return self.EMPTY_ANSWERS
+    end
+
+    processed_answers[1] = cname_answer
+
+    log(DEBUG, "processed cname:", cname_answer.cname)
+
+  else
+    log(DEBUG, "processed ans:", #processed_answers)
+  end
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
+
+  return processed_answers
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+  local key = name .. ":" .. qtype
+  stats_increment(self.stats, key, "query")
+
+  local r, err = resolver:new(self.r_opts)
+  if not r then
+    return nil, "failed to instantiate the resolver: " .. err
+  end
+
+  local start_time = now()
+
+  local options = { additional_section = true, qtype = qtype }
+  local answers, err = r:query(name, options)
+  r:destroy()
+
+  local query_time = now() - start_time -- the time taken for the DNS query
+  local time_str = ("%.3f %.3f"):format(start_time, query_time)
+
+  stats_set_count(self.stats, key, "query_last_time", time_str)
+
+  log(DEBUG, "r:query() ans:", answers and #answers or "-", " t:", time_str)
+
+  if not answers then
+    stats_increment(self.stats, key, "query_fail_nameserver")
+    err = err or "unknown"
+    return nil, "DNS server error: " .. err .. ", Query Time: " .. time_str
+  end
+
+  answers = process_answers(self, name, qtype, answers)
+
+  stats_increment(self.stats, key, answers.errstr and "query_fail:" .. answers.errstr
+                                                   or "query_succ")
+
+  return answers, nil, answers.ttl
+end
+
+
+local function stale_update_task(premature, self, key, name, qtype, short_key)
+  if premature then
+    return
+  end
+
+  local answers = resolve_query(self, name, qtype, {})
```

Comment:
Do we need to set `tries` to a temporary table `{}`? Could it be a simple `nil`?

Reply:
done https://github.com/Kong/kong/pull/12305/commits/b87cc14ab796307cf88d07c7aed08475a9bb0497